Fabulous.XamarinForms/src/Fabulous.XamarinForms.Core/ViewUpdaters.fs

```diff
@@ -38,31 +38,44 @@ module ViewUpdaters =
            (create: 'T -> 'TargetT)
            (attach: 'T voption -> 'T -> 'TargetT -> unit) // adjust attached properties
            (canReuse : 'T -> 'T -> bool) // Used to check if reuse is possible
+           (getKey:'T->string)
            (update: 'T -> 'T -> 'TargetT -> unit) // Incremental element-wise update, only if element reuse is allowed
         =
         match prevCollOpt, collOpt with 
         | ValueSome prevColl, ValueSome newColl when identical prevColl newColl -> ()
         | _, ValueNone -> targetColl.Clear()
         | _, ValueSome coll ->
-            if (coll = null || coll.Length = 0) then
-                targetColl.Clear()
-            else
+            
+
+                let prevColl =
+                    match prevCollOpt with
+                    | ValueSome x -> x
+                    | _ -> [||]
+    
                 // Remove the excess targetColl
-                while (targetColl.Count > coll.Length) do
-                    targetColl.RemoveAt (targetColl.Count - 1)
+
 
                 // Count the existing targetColl
                 // Unused variable n' introduced as a temporary workaround for https://github.com/fsprojects/Fabulous/issues/343
                 let _ = targetColl.Count
                 let n = targetColl.Count
-
+                let targetControls = Dictionary<string,'TargetT>()
+                for i in 0..prevColl.Length-1 do
+                        let key = getKey prevColl.[i]
+                        targetControls.[key]<-targetColl.[i]
+                         
+                while (targetColl.Count > coll.Length) do
+                  targetColl.RemoveAt (targetColl.Count - 1)
+               // targetColl.Clear()                    
+                    
                 // Adjust the existing targetColl and create the new targetColl
                 for i in 0 .. coll.Length-1 do
                     let newChild = coll.[i]
+                    let key = getKey newChild
                     let prevChildOpt = match prevCollOpt with ValueNone -> ValueNone | ValueSome coll when i < n -> ValueSome coll.[i] | _ -> ValueNone
                     let prevChildOpt, targetChild = 
                         if (match prevChildOpt with ValueNone -> true | ValueSome prevChild -> not (identical prevChild newChild)) then
-                            let mustCreate = (i >= n || match prevChildOpt with ValueNone -> true | ValueSome prevChild -> not (canReuse prevChild newChild))
+                            let mustCreate = ((not <| targetControls.ContainsKey(key)) || match prevChildOpt with ValueNone -> true | ValueSome prevChild -> not (canReuse prevChild newChild))
```

Comment:
This coupled with the fact you generate a random key for all controls effectively prevents reuse for controls with no user-defined key.

Reply:
I am agree this condition need to be more complex. Right now we have a problem - how to determine, which control we should cache, and which we should recreate. Unfortunately, we can not instantiate two "instance" of same view element on different view cycles.

Reply:
i>=n is not always valid for support old style in new paradigm. It would be more likely like this: 
 (i>=n && (not <| targetControls.Contains key))

Reply:
I agree. The case where the control has no matching key or has no key should be handled differently. We shouldn't rely on index anymore in my opinion.
See my notes in the first message.

Reply:
For now i am following this way: 
1. I check if  i have control in dictionary. If yes, i apply new logic. If no, old logic takes place

Reply:
I fear that the old logic can't really be kept as-is with randomly placed keyed controls.
The previous logic was only working since all controls were check based on the index indiscriminately.